### PR TITLE
server: add support for log rotation

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds the option to have  `lumberjack`[1] manage log rotation for file paths defined in `log-output`. The flag `--log-rotation-config-json` accepts a JSON config directly reducing the footprint for enabling the feature. Currently, the implementation supports log rotation of a single file target as outlined in the example below. A similar approach is taken by kube-apiserver for audit logs.[2]

new flags
```
  --enable-log-rotation
  --log-rotation-config-json
```

example usage

```
etcd \ 
  --logger zap \
  --log-outputs stderr,/var/log/etcd/etcd.log \
  --enable-log-rotation \
  --log-rotation-config-json '{"maxsize": 1, "maxbackups": 5, "compress": true}'
```

```
-rw-r--r-- 1 root root  21K May  6 20:52 /var/log/etcd/etcd-2021-05-07T00-52-07.344.log.gz
-rw-r--r-- 1 root root  11K May  6 20:52 /var/log/etcd/etcd-2021-05-07T00-52-14.004.log.gz
-rw-r--r-- 1 root root 9.8K May  6 20:52 /var/log/etcd/etcd-2021-05-07T00-52-20.523.log.gz
-rw-r--r-- 1 root root  10K May  6 20:52 /var/log/etcd/etcd-2021-05-07T00-52-26.902.log.gz
-rw-r--r-- 1 root root  11K May  6 20:52 /var/log/etcd/etcd-2021-05-07T00-52-34.252.log.gz
-rw-r--r-- 1 root root 323K May  6 20:52 /var/log/etcd/etcd.log
```
[1] http://github.com/natefinch/lumberjack/
[2] https://github.com/kubernetes/apiserver/blob/fa8a40ce7f57e4ffeb1b811dca2dbed9209d7795/pkg/server/options/audit.go#L27

### CODE REVIEW

File: server/embed/config_logging.go [hunk 1]
```diff
@@ -78,7 +80,19 @@ func (cfg *Config) setupLogging() error {
 				}
 			}
 		}
-
+		if cfg.LogRotate {
+			zap.RegisterSink("rotate", func(u *url.URL) (zap.Sink, error) {
+				return rotationSink{
+					Logger: &lumberjack.Logger{
```
Comment (ptabor) [inline server/embed/config_logging.go:86]: As the lumberjeck config:  is JSON marshallable: 
  https://github.com/natefinch/lumberjack/blob/47ffae23317c5951a2a6267a069cf676edf53eb6/lumberjack.go#L83,

maybe instead being opinionated about which flags to we support in etcd, we should just have a single:
  "--log_rotation_lumberjack_json={...}"

Comment (hexfusion) [reply server/embed/config_logging.go:86]: Yeah I like that +1

**Review by ptabor — CHANGES_REQUESTED** [review]
As discussed in the other comment, marking as 'Request changes' due to 
"--log_rotation_lumberjack_json={...}" idea.

**Comment by hexfusion** [comment]
@ptabor I will update this and tests Friday for review.

**Comment by hexfusion** [comment]
cc @lillc

**Comment by hexfusion** [comment]
@lilic

File: server/embed/config_logging.go [hunk 2]
```diff
@@ -55,6 +57,34 @@ func (cfg *Config) setupLogging() error {
 			}
 		}
 
+		var logRotateConfig *lumberjack.Logger
```
Comment (ptabor) [inline server/embed/config_logging.go:60]: This function is already huge. 
Please consider moving it to a separate method.

File: server/embed/config_logging.go [hunk 3]
```diff
@@ -55,6 +57,34 @@ func (cfg *Config) setupLogging() error {
...
+		if cfg.LogRotateConfigJSON != "" {
+			var logOutputFilePaths int
```
Comment (ptabor) [inline server/embed/config_logging.go:62]: `logOutputFilePathsCnt := 0 ` 

would by IMHO more explicit

File: server/embed/config_logging.go [hunk 4]
```diff
@@ -55,6 +57,34 @@ func (cfg *Config) setupLogging() error {
...
+			}
+			if len(cfg.LogOutputs) == 1 && logOutputFilePaths == 0 {
+				return ErrLogRotationInvalidLogOutput
+			}
+			if logOutputFilePaths > 1 {
+				return ErrLogRotationInvalidLogOutput
+			}
+
+			if err := json.Unmarshal([]byte(cfg.LogRotateConfigJSON), &logRotateConfig); err != nil {
+				return fmt.Errorf("failed to deserialize lumberjack config: %v", cfg.LogRotateConfigJSON)
```
Comment (ptabor) [inline server/embed/config_logging.go:79]: Please provide also original error using '%w' (please see: https://blog.golang.org/go1.13-errors)

File: server/embed/config_test.go [hunk 5]
```diff
@@ -289,3 +289,64 @@ func TestPeerURLsMapAndTokenFromSRV(t *testing.T) {
...
+		name              string
+		logOutputs        []string
+		logRotationConfig string
+		wantErr           bool
+		wantErrMsg        error
+	}{
+		{
+			name:              "mixed log output targets",
+			logOutputs:        []string{"stderr", "/tmp/path"},
+			logRotationConfig: `{"maxsize": 1}`,
```
Comment (ptabor) [inline server/embed/config_test.go:304]: What's the behavior of simplest default config: 

```
 --logRotationConfig='{}'
```
? Maybe worth testing / documenting as part of flag description.

**Review by ptabor — APPROVED** [review]

File: server/embed/config_logging.go (hunk 3 above)
Comment (hexfusion) [reply server/embed/config_logging.go:62]: fair

**Comment by xiang90** [comment]
@hexfusion @ptabor 

How stable is `https://github.com/natefinch/lumberjack`? Do we want a tight-coupled config with that implementation? Or shall we provide something simpler first (or at least not exposed the rotation implementation?)?

**Comment by xiang90** [comment]
@ptabor Does k8s support any built-in log rotation?

**Comment by hexfusion** [comment]
@xiang90 lumberjack is used by kube-apiserver for audit logs[1]. That was my reason for choosing it and the direct reference to its use case in zap FAQ[2].

[1] https://github.com/kubernetes/apiserver/blob/fa8a40ce7f57e4ffeb1b811dca2dbed9209d7795/pkg/server/options/audit.go#L27
[2] https://github.com/uber-go/zap/blob/master/FAQ.md#does-zap-support-log-rotation

**Comment by hexfusion** [comment]
It was implemented in KAS almost 4 years ago so I feel like this should be pretty stable. But I am open to alternatives of course.

[1] https://github.com/kubernetes/apiserver/commit/d3c1c03062ae2cfd3e6c112ff0a89253ca74ba02#diff-a12a91ffb3cb7f3f6b159e4f146b64f72e59667fca475ec3f74e2b6039201d34R21

File: server/etcdmain/config.go [hunk 6]
```diff
@@ -236,6 +236,7 @@ func newConfig() *config {
 	fs.StringVar(&cfg.ec.Logger, "logger", "zap", "Currently only supports 'zap' for structured logging.")
 	fs.Var(flags.NewUniqueStringsValue(embed.DefaultLogOutput), "log-outputs", "Specify 'stdout' or 'stderr' to skip journald logging even when running under systemd, or list of comma separated output targets.")
 	fs.StringVar(&cfg.ec.LogLevel, "log-level", logutil.DefaultLogLevel, "Configures log level. Only supports debug, info, warn, error, panic, or fatal. Default 'info'.")
+	fs.StringVar(&cfg.ec.LogRotateConfigJSON, "log-rotate-config-json", "", "Configures log rotation with a JSON lumberjack logger config. Supports a single file output target.")
```
Comment (xiang90) [inline server/etcdmain/config.go:239]: Can we at least define/re-define the JSON format in the etcd documentation? Our users should not need to learn `lumberjack`. Also, shall we provide a default config? Probably we have to divide the flag into two parts, one to enable the rotation and the other one to config it.

Comment (hexfusion) [reply server/etcdmain/config.go:239]: this makes sense.

**Comment by ptabor** [comment]
@serathius has a lot of experience with observability aspects of k8s. PTAL

**Comment by hexfusion** [comment]
Updated to reflect previous comments will rebase when review is complete

File: server/embed/config.go [hunk 7]
```diff
@@ -68,6 +68,15 @@ const (
 	StdErrLogOutput  = "stderr"
 	StdOutLogOutput  = "stdout"
 
+	// DefaultLogRotationConfig is the default configuration used for log rotation.
+	// Log rotation is disabled by default.
+	// MaxSize    = 100 // MB
+	// MaxAge     = 0 // none
+	// MaxBackups = 0 // no limit
+	// LocalTime  = false // UTC by default
+	// Compress   = false
```
Comment (xiang90) [inline server/embed/config.go:77]: add a comment for this as well? Compress the rotated log in gzip format.

File: server/etcdmain/config.go [hunk 8]
```diff
@@ -236,6 +236,8 @@ func newConfig() *config {
 	fs.StringVar(&cfg.ec.Logger, "logger", "zap", "Currently only supports 'zap' for structured logging.")
 	fs.Var(flags.NewUniqueStringsValue(embed.DefaultLogOutput), "log-outputs", "Specify 'stdout' or 'stderr' to skip journald logging even when running under systemd, or list of comma separated output targets.")
 	fs.StringVar(&cfg.ec.LogLevel, "log-level", logutil.DefaultLogLevel, "Configures log level. Only supports debug, info, warn, error, panic, or fatal. Default 'info'.")
+	fs.BoolVar(&cfg.ec.EnableLogRotation, "enable-log-rotation", false, "Enable log rotation of a single log-outputs file target.")
+	fs.StringVar(&cfg.ec.LogRotationConfigJSON, "log-rotation-config-json", embed.DefaultLogRotationConfig, "Configures log rotation if enabled with a JSON logger config. Default: MaxSize=100, MaxAge=0, MaxBackups=0, LocalTime=false, Compress=false")
```
Comment (xiang90) [inline server/etcdmain/config.go:240]: Add the unit to the example config? For example: MaxSize = 100 (MB), MaxAge = 0 (days).

File: server/etcdmain/help.go [hunk 9]
```diff
@@ -188,6 +188,10 @@ Logging:
     Specify 'stdout' or 'stderr' to skip journald logging even when running under systemd, or list of comma separated output targets.
   --log-level 'info'
     Configures log level. Only supports debug, info, warn, error, panic, or fatal.
+  --enable-log-rotation 'false'
+    Enable log rotation of a single log-outputs file target.
+  --log-rotation-config-json '{"maxsize": 100, "maxage": 0, "maxbackups": 0, "localtime": false, "compress": false}'
```
Comment (xiang90) [inline server/etcdmain/help.go:193]: document the unit of the config? size in MB for example?

**Comment by xiang90** [comment]
LGTM. defer to @ptabor.

File: server/etcdmain/help.go (hunk 9 above)
Comment (hexfusion) [reply server/etcdmain/help.go:193]: will followup on comments

**Comment by ptabor** [comment]
Looks good to me. Thank you.

**Pull request merged**